packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,163 @@
+  async getContractStorage(address: Address, key: Buffer): Promise<Buffer> {
+    const slotCacheKey = `${address.toString()}--0x${key.toString('hex')}`
+    // Check storage slot in cache
+    let storage: Buffer | string | undefined = this.storageCache.get(slotCacheKey)
+    if (typeof storage !== 'undefined') return storage
+    // Retrieve storage slot from provider if not found in cache
+    storage = await this.provider.getStorageAt(address.toString(), bufferToBigInt(key))
+    const value = toBuffer(storage)
+    // Cache retrieved storage slot
+    await this.putContractStorage(address, key, value)
+    return value
+  }
+
+  async putContractStorage(address: Address, key: Buffer, value: Buffer): Promise<void> {
+    // Set value in storageCache with `[address]--[hexKey]` formatted key
+    this.storageCache.set(`${address.toString()}--${bufferToHex(key)}`, value)
+  }
+
+  copy(): EthersStateManager {
+    return new EthersStateManager({ provider: this.provider })
+  }
+
+  dumpStorage(_address: Address): Promise<StorageDump> {
+    throw new Error('Method not implemented.')
+  }
+
+  async accountExists(address: Address): Promise<boolean> {
+    log(`Verify if ${address.toString()} exists`)
+    const account = this._cache.get(address)
+    if (account.isEmpty()) {
+      // Get latest block (or block specified in `this.blockTag`)
+      const block = await this.provider.send('eth_getBlockByNumber', [this.blockTag, false])
+      // Get merkle proof for `address` from provider
+      const proof = await this.provider.send('eth_getProof', [address.toString(), [], block.number])
+      const proofBuf = proof.accountProof.map((proofNode: string) => toBuffer(proofNode))
+      const trie = new Trie({ useKeyHashing: true })
+      const verified = await trie.verifyProof(toBuffer(block.stateRoot), address.buf, proofBuf)
+      // if not verified (i.e. verifyProof returns null), account does not exist
+      if (verified === null) return false
+    }
+    return true
+  }
+
+  async getAccount(address: Address): Promise<Account> {
+    const account = await this._cache.getOrLoad(address)
+    return account
+  }
+  async getAccountFromProvider(address: Address): Promise<Account> {
+    log(`Retrieving account data for ${address.toString()} from provider`)
+
+    const accountData = await this.provider.send('eth_getProof', [
+      address.toString(),
+      [],
+      this.blockTag,
+    ])
+    const rawData = accountData.accountProof
+    for (const proofItem of rawData) {
+      // Dump raw proof nodes to DB under their hashed keys
+      const dataBuffer = Buffer.from(proofItem.slice(2), 'hex')
+      const hash = keccak256(dataBuffer)
+      await (this.trie as any)._db.put(arrToBufArr(hash), dataBuffer)
+    }
+
+    const account = Account.fromAccountData({
+      balance: BigInt(accountData.balance),
+      nonce: BigInt(accountData.nonce),
+      codeHash: toBuffer(accountData.codeHash),
+      storageRoot: await this.getStateRoot(),
```

Comment:
This is not right, the stateRoot of the account should be taken from the accountData. The state root of the account is the root of the underlying trie DB which are the key/value pairs of the storage.